### PR TITLE
Ofri climber example

### PR DESCRIPTION
**Pull request by OfriRosenbaum**

### CODE REVIEW

File: src/main/java/frc/robot/subsystems/climber/Climber.java [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+     * The speed in which the arms will move up to catch the bar (in order to climb).
+     *
+     * <p>A game that this example subsystem can be used in is 2020 FRC INFINITE RECHARGE</p>
+     */
+    public final Supplier<Double> UP_SPEED = rootNamespace.addConstantDouble("up speed", 0.3);
+
+    /**
+     * The speed in which the arms will move down in order to take the robot up after they caught the bar.
+     */
+    public final Supplier<Double> DOWN_SPEED = rootNamespace.addConstantDouble("down speed", -0.4);
```
Comment (tal-sitton) [inline src/main/java/frc/robot/subsystems/climber/Climber.java]: you don't need `MAX_SPEED` and `MIN_SPEED` if you have `UP_SPEED` and `DOWN_SPEED`

**Review by tal-sitton — CHANGES_REQUESTED** [review]

File: src/main/java/frc/robot/subsystems/climber/Climber.java [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+import edu.wpi.first.wpilibj.shuffleboard.Shuffleboard;
+
+import java.util.function.Supplier;
+
+public class Climber extends MotoredGenericSubsystem {
+
+    /**
+     * The speed in which the arms will move up to catch the bar (in order to climb).
+     *
+     * <p>A game that this example subsystem can be used in is 2020 FRC INFINITE RECHARGE</p>
```
Comment (tal-sitton) [inline src/main/java/frc/robot/subsystems/climber/Climber.java]: you know that this year (2022) we also have to climb... right?

File: src/main/java/frc/robot/subsystems/climber/Climber.java [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+import com.ctre.phoenix.motorcontrol.can.WPI_VictorSPX;
+import com.spikes2212.command.genericsubsystem.MotoredGenericSubsystem;
+import com.spikes2212.command.genericsubsystem.commands.MoveGenericSubsystem;
+import edu.wpi.first.wpilibj.shuffleboard.Shuffleboard;
+
+import java.util.function.Supplier;
+
+public class Climber extends MotoredGenericSubsystem {
+
+    /**
```
Comment (tal-sitton) [inline src/main/java/frc/robot/subsystems/climber/Climber.java:18]: Javadoc of a class should be before `public class....`

File: src/main/java/frc/robot/subsystems/climber/Climber.java [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+            WPI_VictorSPX motor = new WPI_VictorSPX(RobotMap.CAN.CLIMBER_VICTOR);
+            motor.setNeutralMode(NeutralMode.Brake);
+            instance = new Climber(motor);
+        }
+        return instance;
+    }
+
+    private WPI_VictorSPX motor;
+
+    public Climber(WPI_VictorSPX motor) {
```
Comment (tal-sitton) [inline src/main/java/frc/robot/subsystems/climber/Climber.java]: why public?

**Review by tal-sitton — CHANGES_REQUESTED** [review]

File: src/main/java/frc/robot/subsystems/climber/Climber.java [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+import com.ctre.phoenix.motorcontrol.NeutralMode;
+import com.ctre.phoenix.motorcontrol.can.WPI_VictorSPX;
+import com.spikes2212.command.genericsubsystem.MotoredGenericSubsystem;
+import com.spikes2212.command.genericsubsystem.commands.MoveGenericSubsystem;
+import edu.wpi.first.wpilibj.shuffleboard.Shuffleboard;
+
+import java.util.function.Supplier;
+
+/**
+ * A game that this example subsystem can be used in is FRC Rapid React 2022.
```
Comment (tal-sitton) [inline src/main/java/frc/robot/subsystems/climber/Climber.java]: explain what is this subsystem

**Review by tal-sitton — CHANGES_REQUESTED** [review]

File: src/main/java/frc/robot/subsystems/climber/Climber.java [hunk 6]
```diff
@@ -0,0 +1,54 @@
...
+    public static final double UP_SPEED = 0.3;
+
+    /**
+     * The speed in which the arms will move down in order to take the robot up after they caught the bar.
+     */
+    public static final double DOWN_SPEED = -0.4;
+
+    private WPI_VictorSPX motor;
+
+    private static Climber instance;
```
Comment (Yoel-Perman-Brilliant) [inline src/main/java/frc/robot/subsystems/climber/Climber.java:34]: explain

**Review by Yoel-Perman-Brilliant — DISMISSED** [review]

File: src/main/java/frc/robot/subsystems/climber/Climber.java [hunk 7]
```diff
@@ -0,0 +1,55 @@
...
+import com.spikes2212.command.genericsubsystem.MotoredGenericSubsystem;
+import com.spikes2212.command.genericsubsystem.commands.MoveGenericSubsystem;
+import edu.wpi.first.wpilibj.shuffleboard.Shuffleboard;
+
+import java.util.function.Supplier;
+
+/**
+ * A subsystem which represents a climber. The climber consists of an arm that is controlled by a VictorSPX. The arm
+ * can move up and grab a bar, then down to pull the robot up; thus making the robot climb. <br>
+ * A game that this example subsystem can be used in is FRC Rapid React 2022.
```
Comment (tal-sitton) [inline src/main/java/frc/robot/subsystems/climber/Climber.java]: why did you write `FRC`? isn't it obvious?

File: src/main/java/frc/robot/subsystems/climber/Climber.java [hunk 8]
```diff
@@ -0,0 +1,55 @@
...
+     */
+    public static final double UP_SPEED = 0.3;
+
+    /**
+     * The speed in which the arms will move down in order to take the robot up after they caught the bar.
+     */
+    public static final double DOWN_SPEED = -0.4;
+
+    /**
+     * With only a private constructor and a single private instance with a public getter, the singleton design pattern
```
Comment (tal-sitton) [inline src/main/java/frc/robot/subsystems/climber/Climber.java]: sounds like a copy-paste from some Shakespeare book. CHANGE

**Review by tal-sitton — CHANGES_REQUESTED** [review]

File: src/main/java/frc/robot/subsystems/climber/Climber.java [hunk 9]
```diff
@@ -0,0 +1,56 @@
...
+import com.ctre.phoenix.motorcontrol.NeutralMode;
+import com.ctre.phoenix.motorcontrol.can.WPI_VictorSPX;
+import com.spikes2212.command.genericsubsystem.MotoredGenericSubsystem;
+import com.spikes2212.command.genericsubsystem.commands.MoveGenericSubsystem;
+import edu.wpi.first.wpilibj.shuffleboard.Shuffleboard;
+
+import java.util.function.Supplier;
+
+/**
+ * A subsystem which represents a climber. The climber consists of an arm that is controlled by a VictorSPX. The arm
```
Comment (tal-sitton) [inline src/main/java/frc/robot/subsystems/climber/Climber.java]: we know this class represents a climber, its in the name

File: src/main/java/frc/robot/subsystems/climber/Climber.java [hunk 10]
```diff
@@ -0,0 +1,56 @@
...
+     */
+    public static final double UP_SPEED = 0.3;
+
+    /**
+     * The speed in which the arms will move down in order to take the robot up after they caught the bar.
+     */
+    public static final double DOWN_SPEED = -0.4;
+
+    /**
+     * The Climber class is a singleton, which means it has only one instance. Since the constructor is private, a new
```
Comment (tal-sitton) [inline src/main/java/frc/robot/subsystems/climber/Climber.java:29]: "But why would you want it to be singleton?" ~ a random programmer

**Review by tal-sitton — CHANGES_REQUESTED** [review]

**Review by tal-sitton — APPROVED** [review]

**Pull request merged**